### ISSUE TITLE
Multi-GPU / Multi-Node Worker Crash / Deadlock When Returning {"loss": None, "lr": None}

### ISSUE THREAD

**Issue by themikem** [NVIDIA/NeMo#3801]
**Describe the bug**

When running multi-GPU / multi-node training of NLP Entity Linking model against full UMLS dataset, a return of None for loss & lr from training_step results in a crash of a worker and deadlock (at 100% GPU / CPU utilization). It appears this is incompatible with multi-GPU, but entity linking notebooks reference multi-gpu training.

**Steps/Code to reproduce bug**

Training fails at around 75% of first epoch, which is ~18 hours in on 4x A100s, or more than double that on single GPU setups.

Error message is "pytorch_lightning.utilities.exceptions.MisconfigurationException: In automatic_optimization, when `training_step` returns a dict, the 'loss' key needs to be present"

Relevant code in training_step: https://github.com/NVIDIA/NeMo/blob/69385daa999907a15f86e461b670a866424f5bb9/nemo/collections/nlp/models/entity_linking/entity_linking_model.py#L83-L103

Note PyTorch Lightning docs on [training_step](https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#training-step) return value:
```
None - Training will skip to the next batch. This is only for automatic optimization.
**This is not supported for multi-GPU, TPU, IPU, or DeepSpeed.**
```

Thus this will work on single GPU, but not multi-GPU according to the docs.

**Expected behavior**

Training completes without crashing.

**Environment overview (please complete the following information)**

 - Environment location: Docker 
 - Method of NeMo install: Nemo NGC v.1.6.1
 - If method of install is [Docker], provide `docker pull` & `docker run` commands used
 Pull: docker pull nvcr.io/nvidia/nemo:1.6.1
Run: docker run --gpus all -it -v /opt/nvidia/nemo:/workspace/nemo/local --shm-size=16g --ulimit memlock=-1 --ulimit stack=67108864 -v /dev/infiniband --uts=host --net=host --ipc=host --cap-add IPC_LOCK --device=/dev/snd --device /dev/infiniband/uverbs0 --device /dev/infiniband/rdma_cm --device /dev/infiniband/umad0 nvcr.io/nvidia/nemo:1.6.1 /bin/bash

**Environment details**

If NVIDIA docker image is used you don't need to specify these.
Otherwise, please provide:
- OS version
- PyTorch version
- Python version

**Additional context**

Tested on A100 & Titan RTX - same results
Single vs. Multi-node does not appear to make a difference either.

**Comment by themikem** [comment]
@vadam5 It looks likes you may be familiar with this part of the code... any thoughts on how to restructure this to either ignore this part of the batch when averaging across multiple GPUs, or otherwise skip the batch successfully when using ddp? Thanks for any insights...

**Comment by VahidooX** [comment]
As you have given the reference, returning None from training_step would not work with multi-gpu and it is a limilation comes from PTL. Why not just return loss zero when you want to drop that batch?

**Comment by themikem** [comment]
After trying this on a single node / GPU (1x A100 card), we have encountered a similar issue after about 5 days of training. Somewhere around global step 75999 (per the lightning_logs.txt) we hit a batch with no hard examples:
`[NeMo I 2022-03-30 17:19:33 multi_similarity_loss:91] Encountered zero loss in multisimloss, loss = 0.0. No hard examples found in the batch`

This then resulted in a crash of the training with this message: 
`pytorch_lightning.utilities.exceptions.MisconfigurationException: In automatic_optimization, when training_step returns a dict, the 'loss' key needs to be present`

Checking the entity_linking_model.py file, we have for training_step:
```
def training_step(self, batch, batch_idx):
        """
        Lightning calls this inside the training loop with the data from the training dataloader
        passed in as 'batch'.
        """
        input_ids, token_type_ids, attention_mask, concept_ids = batch
        logits = self.forward(input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask)
        train_loss = self.loss(logits=logits, labels=concept_ids)

        # No hard examples found in batch,
        # shouldn't use this batch to update model weights
        if train_loss == 0:
            train_loss = None
            lr = None

        else:
            lr = self._optimizer.param_groups[0]["lr"]
            self.log("train_loss", train_loss)
            self.log("lr", lr, prog_bar=True)

        return {"loss": train_loss, "lr": lr}
```

So it appears that something is filtering the return or otherwise no longer successfully dropping the batch when loss == None...

In either case, it appears there's a conflict between the loss module as implemented and the current state of Nemo / PTL. Appreciate any ideas or suggestions on how to resolve this...

**Comment by themikem** [comment]
> As you have given the reference, returning None from training_step would not work with multi-gpu and it is a limitation comes from PTL. Why not just return loss zero when you want to drop that batch?

Also note that I have tried a few versions of this, but so far none completely successful. It seems that if you hit a zero return from loss or val_loss from a single node in the batch just right, it still deadlocks, even with 4 GPUs in the same physical machine. (i.e. isn't a network problem between nodes).  Still attempting to determine the root cause, but it isn't quite as simple as just pushing a zero value through it seems.

**Comment by titu1994** [comment]
If 0 occurs on only one GPU but other 3 GPUs have values, then nccl will hang cause it will forever wait for the 1 GPU with 0 result. You need to do a global search across all ranks for such cases, and skip updates on every single rank for that step.

**Comment by themikem** [comment]
@titu1994 Thank you! This confirms what I have seen across tests on three different architectural attempts. I've been focused on PTL - didn't occur to me to look at nccl for this.

Do you think it is sufficient to implement this in training_step_end (and possibly switch to ddp2), or would you expect the nccl issue has already occurred by the time ptl hits that call? (I need to do more research to understand the precise behavior with nccl here...)

My other thought, while less than ideal, was to code "0" as 1e-38 or some such so that it is not explicitly zero but probably mathematically equivalent in practical terms. And seeing as I already don't want to be passing zeros in the first place, that might be better in the sense that it can complete training. Obviously then would like to find a better solution in the long term...

Thanks again for confirming this!

**Comment by titu1994** [comment]
So you must to a global test and collate the results into a single tensor via torch distributed all_reduce() call to check if any rank has 0, and only then perform the if value == 0 check. If any value has 0, then you need to 0 out loss and submit 0 loss to every single rank (or if PTL provides a way to skip a step, do that instead).

**Comment by titu1994** [comment]
For an example you can look at ConvASREncoder or ConformerEncoder - they have blocks to compute the global max sequence length and then set that as the new buffer across all rank for exactly this reason.

**Comment by themikem** [comment]
> So you must to a global test and collate the results into a single tensor via torch distributed all_reduce() call to check if any rank has 0, and only then perform the if value == 0 check. If any value has 0, then you need to 0 out loss and submit 0 loss to every single rank (or if PTL provides a way to skip a step, do that instead).

Okay got it - the all_reduce call was the basis of my question on the training_step_end - i.e. was not certain if that is where nccl was locking up because of the included zero. This sounds more related to what goes on between autograd, the backwards call, and nccl. Regardless I'll dig into the examples you cited and then ping you back if I get stuck. 

My understanding from some other PTL documents is that an explicit None for the loss implies "skip batch", but I haven't actually seen it get far enough to directly observe that. There was also a patch buried in the PTL pull requests to handle all this at the framework level but it doesn't look like it was ever completed.

**Comment by titu1994** [comment]
Yeah we've been bitten by some issues recently too. It happens with a major PTL release, usually ironed out soon but this would be better to solve at code level in your model itself to be robust to multi GPU training.

**Comment by sreeramjoopudi** [comment]
I ran into the same issue and I appreciate the above comments mentioned above as it gives me an idea on how to fix the issue.
 @themikem : As this is closed, I wanted to double-check if the fix has been delivered (I couldn't find the usage of  torch.distributed.all_reduce in https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/entity_linking/entity_linking_model.py#L86). If you're planning to add the fix, it would be of great help if you can update this issue once the fix has been added.
